File: java/tcp_client_sum/TcpClientSum.java
import java.io.*;
import java.net.*;

/**
 * Represents an TCP client.
 */
class TcpClientSum {

    public static final int SERVER_PORT = 9999;

    public static final int SERVER_ADDRESS_REQUIRED = -1;

    public static final int INVALID_SERVER_ADDRESS = -2;

    public static final int TCP_CONNECTION_FAILED = -3;

    /**
     * Connects to a given server and sends TCP messages.
     *
     * @param args server IP address or DNS
     * @throws IOException Input/Output exception
     */
    public static void main(String args[]) throws IOException {
        long f, i, n, num;
        String sentence;
        Socket sock = null;
        InetAddress serverAddress = null;

        if (args.length != 1) {
            System.out.println("Server IPv4/IPv6 address or DNS name is required as argument");
            System.exit(SERVER_ADDRESS_REQUIRED);
        }

        try {
            serverAddress = InetAddress.getByName(args[0]);
        } catch (UnknownHostException ex) {
            System.out.println("Invalid server address supplied: " + args[0]);
            System.exit(INVALID_SERVER_ADDRESS);
        }

        try {
            sock = new Socket(serverAddress, SERVER_PORT);
        }
        catch(IOException ex) {
            System.out.println("Failed to establish TCP connection");
            System.exit(TCP_CONNECTION_FAILED);
        }

        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        DataOutputStream sOut = new DataOutputStream(sock.getOutputStream());
        DataInputStream sIn = new DataInputStream(sock.getInputStream());

        do {
            do {
                num = -1;
                while(num < 0) {
                    System.out.print("Enter a positive integer to SUM (zero to terminate): ");
                    sentence = in.readLine();

                    try {
                        num = Integer.parseInt(sentence);
                    }
                    catch(NumberFormatException ex) {
                        num = -1;
                    }

                    if(num < 0) {
                        System.out.println("Invalid number");
                    }
                }

                n = num;
                for(i = 0; i < 4; i++) {
                    sOut.write((byte)(n % 256));
                    n = n / 256;
                }
            }
            while(num != 0);
            num = 0;
            f = 1;
            for(i = 0; i < 4; i++) {
                num = num + f * sIn.read();
                f = f * 256;
            }
            System.out.println("SUM RESULT = " + num);
        }
        while(num != 0);

        sock.close();
    }
}

File: c/tcp_client_sum/tcp_client_sum.c
/**
 * tcp_client_sum.c
 */

#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>

#include "constants.h"

#define BUFFER_SIZE 30
#define SERVER_PORT "9999"

#define GETS(B,S) {fgets(B, S, stdin); B[strlen(B)-1] = 0;}

int main(int argc, char **argv)
{
    int err, sock;
    unsigned long f, i, n, num;
    unsigned char bt;
    char linha[BUFFER_SIZE];
    struct addrinfo req, *list;

    if(argc != 2) {
        puts("Server IPv4/IPv6 address or DNS name is required as argument");
        exit(SERVER_ADDRESS_REQUIRED);
    }

    bzero((char *)&req, sizeof(req));
    // let getaddrinfo set the family depending on the supplied server address
    req.ai_family = AF_UNSPEC;
    req.ai_socktype = SOCK_STREAM;
    err = getaddrinfo(argv[1], SERVER_PORT, &req, &list);
    if(err != 0) {
        printf("Failed to get server address, error: %s\n", gai_strerror(err));
        exit(INVALID_SERVER_ADDRESS);
    }

    sock = socket(list->ai_family, list->ai_socktype, list->ai_protocol);
    if(sock < 0) {
        perror("Failed to open socket");
        freeaddrinfo(list);
        exit(OPEN_SOCKET_FAILED);
    }

    if(connect(sock, (struct sockaddr *)list->ai_addr, list->ai_addrlen) < 0) {
        perror("Failed connect");
        freeaddrinfo(list);
        close(sock);
        exit(CONNECTION_FAILED);
    }

    do {
        do {
            printf("Enter a positive integer to SUM (zero to terminate): ");
            GETS(linha, BUFFER_SIZE);
            while(sscanf(linha, "%li", &num) != 1 || num < 0)
            {
                printf("Invalid number, please enter a positive integer to "
                     "SUM (zero to terminate): ");
                GETS(linha, BUFFER_SIZE);
            }

            n = num;
            for(i = 0; i < 4; i++)
            {
                bt = n % 256;
                write(sock, &bt, 1);
                n = n / 256;
            }
        }
        while(num > 0);

        num = 0;
        f = 1;
        for(i = 0; i < 4; i++) {
            read(sock, &bt, 1);
            num = num + bt * f;
            f = f * 256;
        }
        printf("SUM RESULT = %lu\n", num);
    }
    while(num > 0);

    close(sock);
    return 0;
}
